### PR TITLE
Add svoice recipe for wsj0-2mix speech separation

### PR DESCRIPTION
**Pull request by nateanl**
- Add [svoice](https://github.com/facebookresearch/svoice) separator proposed in [Single Channel Voice Separation for Unknown Number of Speakers Under Reverberant and Noisy Settings](https://arxiv.org/pdf/2011.02329.pdf)
- Add multi-layer PIT wrapper that weighted sum the SI-SNR loss of multiple RNN layers' outputs.
- Add configuration file for wsj0-2mix experiment with svoice.

### CODE REVIEW

**Comment by sw005320** [comment]
- Can you merge the latest master?
- Can you add the result and model link to https://github.com/espnet/espnet/blob/master/egs2/wsj0_2mix/enh1/README.md?

**Comment by nateanl** [comment]
Sure. The model is under training, I will update the result table and upload the model once the training is finished.

**Comment by nateanl** [comment]
I uploaded the model to https://huggingface.co/Zhaoheng/svoice_wsj0_2mix

**Comment by sw005320** [comment]
@Emrys365, can you review this PR?

@nateanl, can you add a test?

**Comment by Emrys365** [comment]
Sure.

File: egs2/wsj0_2mix/enh1/README.md [hunk 1]
```diff
@@ -91,3 +91,25 @@ config: conf/tuning/train_enh_dc_crn_mapping_snr.yaml
...
+
+## svoice (multi-layer Si-SNR loss)
+
+ - config: conf/tuning/train_enh_svoice.yaml
+ - Pretrained model: https://huggingface.co/Zhaoheng/svoice_wsj0_2mix
+
+|dataset|STOI|SAR|SDR|SIR|
+|---|---|---|---|---|
+|enhanced_cv_min_8k|0.97|21.44|20.98|32.21|
+|enhanced_tt_min_8k|0.98|21.41|20.96|32.27|
```
Comment (Emrys365) [inline egs2/wsj0_2mix/enh1/README.md:114]: Could you also add the SI-SNR results? I think they are already there after running stage 8. You could obtain the scores by running:
```bash
paste /path/to/espnet/egs2/wsj0_2mix/enh1/exp/your_exp_name/enhanced_cv_min_8k/scoring/SI_SNR_spk{1,2} | \
    awk 'BEGIN{sum=0}{n=0;score=0;for (i=2; i<=NF; i+=2){n+=1;score+=$i}; sum+=score/n}END{printf ("SI_SNR(cv)=%.2f\n",sum/NR)}'
```
and
```bash
paste /path/to/espnet/egs2/wsj0_2mix/enh1/exp/your_exp_name/enhanced_tt_min_8k/scoring/SI_SNR_spk{1,2} | \
    awk 'BEGIN{sum=0}{n=0;score=0;for (i=2; i<=NF; i+=2){n+=1;score+=$i}; sum+=score/n}END{printf ("SI_SNR(tt)=%.2f\n",sum/NR)}'
```

File: espnet2/enh/layers/dpmulcat.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        dropout: float = 0.0,
+        bidirectional: bool = True,
+    ):
+        super(MulCatBlock, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/layers/dpmulcat.py:24]:
```suggestion
        super().__init__()
```

File: espnet2/enh/layers/dpmulcat.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            num_layers=1,
+            batch_first=True,
+            dropout=dropout,
+            bidirectional=bidirectional,
+        )
+        self.gate_rnn_proj = nn.Linear(hidden_size * num_direction, input_size)
+
+        self.block_projection = nn.Linear(input_size * 2, input_size)
+
+    def forward(self, input):
```
Comment (Emrys365) [inline espnet2/enh/layers/dpmulcat.py:50]: Could you add a docstring to describe the expected input and output shapes? It would greatly help in understanding how to use this module.

File: espnet2/enh/layers/dpmulcat.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            else:
+                # used to disable normalization
+                self.rows_normalization.append(nn.Identity())
+                self.cols_normalization.append(nn.Identity())
+
+        self.output = nn.Sequential(
+            nn.PReLU(), nn.Conv2d(input_size, output_size * num_spk, 1)
+        )
+
+    def forward(self, input):
```
Comment (Emrys365) [inline espnet2/enh/layers/dpmulcat.py:137]: ditto

File: espnet2/enh/layers/dpmulcat.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        input_size: int,
+        hidden_size: int,
+        output_size: int,
+        num_spk: int,
+        dropout: float = 0.0,
+        num_layers: int = 4,
+        bidirectional: bool = True,
+        input_normalize: bool = False,
+    ):
+        super(DPMulCat, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/layers/dpmulcat.py:106]:
```suggestion
        super().__init__()
```

File: espnet2/enh/loss/wrappers/multilayer_pit_solver.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+from espnet2.enh.loss.wrappers.pit_solver import PITSolver
+
+
+class MultiLayerPITSolver(AbsLossWrapper):
+    def __init__(
+        self,
+        criterion: AbsEnhLoss,
+        weight=1.0,
+        independent_perm=True,
+    ):
```
Comment (Emrys365) [inline espnet2/enh/loss/wrappers/multilayer_pit_solver.py:14]:
```suggestion
    ):
        """Multi-Layer Permutation Invariant Training Solver.
        
        <It would be great if you could add a few words to describe the multi-layer PIT>

        Args:
            criterion (AbsEnhLoss): an instance of AbsEnhLoss
            weight (float): weight (between 0 and 1) of current loss
                for multi-task learning.
            independent_perm (bool):
                If True, PIT will be performed in forward to find the best permutation;
                If False, the permutation from the last LossWrapper output will be
                inherited.
                NOTE (wangyou): You should be careful about the ordering of loss
                    wrappers defined in the yaml config, if this argument is False.
        """
```

File: espnet2/enh/separator/svoice_separator.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
+from collections import OrderedDict
+from typing import List
+from typing import Tuple
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from espnet2.enh.layers.dpmulcat import DPMulCat
+from espnet2.enh.layers.dprnn import split_feature, merge_feature
+from espnet2.enh.separator.abs_separator import AbsSeparator
```
Comment (Emrys365) [inline espnet2/enh/separator/svoice_separator.py:10]:
```suggestion
from collections import OrderedDict
import math
from typing import List
from typing import Tuple

import torch
import torch.nn as nn
import torch.nn.functional as F

from espnet2.enh.layers.dpmulcat import DPMulCat
from espnet2.enh.layers.dprnn import merge_feature
from espnet2.enh.layers.dprnn import split_feature
from espnet2.enh.separator.abs_separator import AbsSeparator
```

File: espnet2/enh/separator/svoice_separator.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            and rank must be at least 2.
+        frame_step: An integer denoting overlap offsets.
+            Must be less than or equal to frame_length.
+    Returns:
+        A Tensor with shape [..., output_size] containing the
+            overlap-added frames of signal's inner-most two dimensions.
+        output_size = (frames - 1) * frame_step + frame_length
+    Based on
+        https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/contrib/signal/python/ops/reconstruction_ops.py
+    """
```
Comment (Emrys365) [inline espnet2/enh/separator/svoice_separator.py:30]:
```suggestion
    """Reconstructs a signal from a framed representation.

    Adds potentially overlapping frames of a signal with shape
    `[..., frames, frame_length]`, offsetting subsequent frames by `frame_step`.
    The resulting tensor has shape `[..., output_size]` where
        output_size = (frames - 1) * frame_step + frame_length

    Args:
        signal: A [..., frames, frame_length] Tensor. All dimensions may be unknown,
            and rank must be at least 2.
        frame_step: An integer denoting overlap offsets.
            Must be less than or equal to frame_length.
    Returns:
        A Tensor with shape [..., output_size] containing the
            overlap-added frames of signal's inner-most two dimensions.
        output_size = (frames - 1) * frame_step + frame_length

    Based on
    https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/contrib/signal/python/ops/reconstruction_ops.py
    """
```

File: espnet2/enh/separator/svoice_separator.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        est_source = nn.AvgPool2d((1, self.kernel_size))(est_source)
+        est_source = overlap_and_add(est_source, self.kernel_size // 2)
+
+        return est_source
+
+
+class SVoiceSeparator(AbsSeparator):
+    """SVoice model for speech separation.
+
+    Args:
```
Comment (Emrys365) [inline espnet2/enh/separator/svoice_separator.py:91]:
```suggestion
    """SVoice model for speech separation.
    
    Reference:
        Voice Separation with an Unknown Number of Multiple Speakers;
        E. Nachmani et al., 2020;
        https://arxiv.org/abs/2003.01531

    Args:
```

**Review by Emrys365 — APPROVED** [review]
LGTM! I just made some minor comments.

**Comment by Emrys365** [comment]
Could you format the python scripts to pass `black` and `flake8` style checks?

Also, could you add some unit tests in [`test/espnet2/enh/test_espnet_model.py`](https://github.com/espnet/espnet/blob/master/test/espnet2/enh/test_espnet_model.py) and `test/espnet2/enh/separator/test_svoice_separator.py`.

> You could take existing unit tests under [`test/espnet2/enh/`](https://github.com/espnet/espnet/blob/master/test/espnet2/enh) as a reference.

File: egs2/wsj0_2mix/enh1/README.md (hunk 1 above)
Comment (nateanl) [reply egs2/wsj0_2mix/enh1/README.md:114]: I see. The [`SI_SNR`](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/enh.sh#L79)  protocol was not there yet when I forked the repo. Let me rerun the scoring script and add the results.

**Comment by sw005320** [comment]
Thanks, @nateanl!
